### PR TITLE
Implement ACP capability masking at protocol proxy seam (2.6.1)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Intercept and mask terminal/* and fs/* ACP capabilities in the protocol proxy path before forwarding to the container. 
- Mask only the first newline-delimited ACP initialize frame; pass through malformed or non-ACP frames unchanged. 
- Preserve line endings and any trailing protocol data after the masked frame. 
- Adds unit and behavior-driven tests, and updates documentation and roadmaps.

### Core functionality
- Updated src/engine/connection/exec/protocol.rs to implement ACP masking at the protocol seam:
  - Added constants: ACP_INITIALIZE_METHOD, ACP_CLIENT_CAPABILITIES_FIELD, ACP_FILE_SYSTEM_CAPABILITY, ACP_TERMINAL_CAPABILITY.
  - Introduced forward_initial_protocol_frame_async to read and forward the first frame with masking.
  - Implemented mask_acp_initialize_frame, parse_json_message, remove_masked_acp_capabilities, and serialize_masked_message to perform safe, minimal rewrites.
  - Preserved bounded buffering and the existing stdout/stderr behavior; masking occurs strictly on the first ACP initialize frame.

### Tests
- Added new unit tests and scaffolding:
  - src/engine/connection/exec/protocol_acp_tests.rs: unit tests for masking logic, line-ending preservation, and trailing bytes.
  - tests/features/acp_capability_masking.feature: behavior-driven tests covering masking, malformed frames, and no-op scenarios.
- The tests exercise: (a) blocked-capability masking, (b) pass-through for malformed/non-ACP frames, and (c) preservation of trailing protocol bytes.

### Tests scaffolding
- Introduced protocol-related ACP tests to verify the narrow seam behavior without impacting the broader protocol proxy.

### Documentation & Roadmap
- docs/podbot-design.md: explicit note on first-frame masking decision and pass-through for malformed frames.
- docs/users-guide.md: describes ACP masking behavior within protocol mode and its lack of user-visible side effects.
- docs/podbot-roadmap.md: marks ACP masking interception as completed for Step 2.6.
- docs/execplans/2-6-1-intercept-acp-initialization.md: new execution plan detailing scope and approach.

### Misc
- No public API changes; ACP masking is contained to protocol.rs.
- No new production dependencies; relies on existing serde_json/Value utilities.

## Why this matters
- Enforces ACP capability masking boundaries early in the protocol path, ensuring hosted agents never learn terminal or filesystem delegation capabilities, in line with design goals. This keeps container sandbox boundaries intact while preserving protocol semantics for non-ACP traffic.

## Validation & Testing
- Gate checks to run locally: make fmt, make lint, and make test; includes the new ACP masking tests.
- Behavior verified for startup/steady-state/shutdown paths via unit tests and BDD scenarios.

## Potential follow-ups
- Extend tests to cover edge cases (empty clientCapabilities, unusual frame terminations).
- Introduce additional regression tests if roadmap gates require broader ACP coverage.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/afef9711-3240-4f69-8b15-8bbb229d1ab0

## Summary by Sourcery

Implement ACP capability masking at the protocol stdin proxy seam while documenting and validating the behaviour with unit, BDD, and roadmap updates.

New Features:
- Mask ACP `initialize` requests in protocol mode to remove `terminal/*` and `fs/*` capabilities before forwarding to the container while preserving non-ACP traffic semantics.

Enhancements:
- Refine protocol proxy stdin handling to inspect and rewrite only the first newline-delimited ACP frame, maintaining bounded buffering and existing shutdown behaviour.
- Document ACP initialization masking and buffering behaviour in the design and user guides, including the first-frame-only rewrite and pass-through of malformed or non-ACP frames.
- Clarify and reflow several existing execution plans and design docs for readability without changing their technical intent.

Documentation:
- Introduce a new ExecPlan describing ACP initialization interception (Step 2.6.1) and mark the corresponding roadmap task as complete.

Tests:
- Add focused unit and BDD-style tests to verify ACP capability masking, handling of malformed or non-ACP frames, and preservation of trailing protocol bytes in the protocol stdin proxy.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="src/engine/connection/exec/protocol_acp_tests.rs" line_range="65-74" />
<code_context>
+    }
+}
+
+fn initialize_frame(line_ending: &str) -> Vec<u8> {
+    let payload = serde_json::json!({
+        "jsonrpc": "2.0",
+        "id": 0,
+        "method": "initialize",
+        "params": {
+            "protocolVersion": 1,
+            "clientCapabilities": {
+                "fs": {
+                    "readTextFile": true,
+                    "writeTextFile": true
+                },
+                "terminal": true,
+                "_meta": {
+                    "custom": true
+                }
+            },
+            "clientInfo": {
+                "name": "podbot-tests",
+                "version": "1.0.0"
+            }
+        }
+    });
+
+    let mut frame = serde_json::to_vec(&payload).expect("initialize payload should serialize");
+    frame.extend_from_slice(line_ending.as_bytes());
+    frame
+}
+
</code_context>
<issue_to_address>
**suggestion (testing):** Add a test for initialize frames where `clientCapabilities` only contains blocked capabilities so the field gets removed

Currently all tests keep a non-blocked `_meta` entry, so `clientCapabilities` never becomes empty and the branch that removes the entire field is untested. Please add a unit test that builds an `initialize` frame whose `clientCapabilities` only includes `fs` and/or `terminal`, and then assert that after masking: (1) `clientCapabilities` is removed, and (2) other `params` fields such as `clientInfo` and `protocolVersion` are unchanged. This will cover the `client_capabilities.is_empty()` path in `remove_masked_acp_capabilities`.

Suggested implementation:

```rust
use rstest::rstest;
use rstest_bdd::Slot;
use rstest_bdd_macros::{ScenarioState, given, scenario, then, when};
use tokio::io::{AsyncWrite, AsyncWriteExt, DuplexStream};

#[test]
fn remove_masked_acp_capabilities_removes_empty_client_capabilities() {
    // Build an initialize payload whose `clientCapabilities` only contains
    // capabilities that should be fully masked (`fs` and `terminal`).
    let mut payload = serde_json::json!({
        "jsonrpc": "2.0",
        "id": 0,
        "method": "initialize",
        "params": {
            "protocolVersion": 1,
            "clientCapabilities": {
                "fs": {
                    "readTextFile": true,
                    "writeTextFile": true
                },
                "terminal": true
            },
            "clientInfo": {
                "name": "podbot-tests",
                "version": "1.0.0"
            }
        }
    });

    // Exercise the masking logic.
    remove_masked_acp_capabilities(&mut payload);

    // `clientCapabilities` should be completely removed once all entries are masked.
    assert!(
        payload["params"].get("clientCapabilities").is_none(),
        "`clientCapabilities` should be removed when all capabilities are masked"
    );

    // Other `params` fields must remain unchanged.
    assert_eq!(
        payload["params"]["protocolVersion"],
        serde_json::json!(1),
        "`protocolVersion` should remain unchanged after masking"
    );

    assert_eq!(
        payload["params"]["clientInfo"],
        serde_json::json!({
            "name": "podbot-tests",
            "version": "1.0.0"
        }),
        "`clientInfo` should remain unchanged after masking"
    );
}

```

1. Ensure that `remove_masked_acp_capabilities` is in scope for this test. If it lives in the same module as other ACP tests, you may need a `use super::remove_masked_acp_capabilities;` near the top of the file.
2. If the real API operates on frames/bytes instead of a `serde_json::Value`, adapt the test body to:
   * Serialize `payload` to a frame (including the correct line ending),
   * Pass that frame through the masking function under test,
   * Deserialize back to `serde_json::Value` before asserting on `params`.
3. If `remove_masked_acp_capabilities` also strips the `protocolVersion` or modifies `clientInfo` in your implementation, adjust the expectations accordingly so they match the intended behavior described in your comment.
</issue_to_address>

### Comment 2
<location path="src/engine/connection/exec/protocol.rs" line_range="209" />
<code_context>
+        .await
+}
+
+fn mask_acp_initialize_frame(frame: &[u8]) -> Vec<u8> {
+    let (payload, line_ending) = split_frame_line_ending(frame);
+    let Some(mut message) = parse_json_message(payload) else {
</code_context>
<issue_to_address>
**issue (complexity):** Consider inlining the serde parsing/serialization calls into `mask_acp_initialize_frame` and removing the extra helper functions to keep the masking logic in one readable place.

You can simplify `mask_acp_initialize_frame` by inlining the thin serde wrappers and reducing the number of tiny helpers. This keeps the masking behavior identical but makes the control flow easier to follow.

For example:

```rust
fn mask_acp_initialize_frame(frame: &[u8]) -> Vec<u8> {
    let (payload, line_ending) = split_frame_line_ending(frame);

    let mut message: Value = match serde_json::from_slice(payload) {
        Ok(v) => v,
        Err(_) => return frame.to_vec(),
    };

    if !remove_masked_acp_capabilities(&mut message) {
        return frame.to_vec();
    }

    let mut serialized = match serde_json::to_vec(&message) {
        Ok(v) => v,
        Err(_) => return frame.to_vec(),
    };
    serialized.extend_from_slice(line_ending);
    serialized
}
```

With this, you can drop `parse_json_message` and `serialize_masked_message` entirely, and keep `remove_masked_acp_capabilities` as the one domain-specific helper:

```rust
fn remove_masked_acp_capabilities(message: &mut Value) -> bool {
    if message.get("method").and_then(Value::as_str) != Some(ACP_INITIALIZE_METHOD) {
        return false;
    }

    let Some(params) = message.get_mut("params").and_then(Value::as_object_mut) else {
        return false;
    };

    let Some(client_capabilities) = params
        .get_mut(ACP_CLIENT_CAPABILITIES_FIELD)
        .and_then(Value::as_object_mut)
    else {
        return false;
    };

    let removed_terminal = client_capabilities.remove(ACP_TERMINAL_CAPABILITY).is_some();
    let removed_fs = client_capabilities.remove(ACP_FILE_SYSTEM_CAPABILITY).is_some();
    if !removed_terminal && !removed_fs {
        return false;
    }

    if client_capabilities.is_empty() {
        params.remove(ACP_CLIENT_CAPABILITIES_FIELD);
    }

    true
}
```

This keeps all fallback behavior (`return frame.to_vec()` on any parse/serialize failure or no-op case) but removes indirection and makes the overall transformation readable in one place.
</issue_to_address>

### Comment 3
<location path="docs/execplans/2-6-1-intercept-acp-initialization.md" line_range="1" />
<code_context>
+# Step 2.6.1: Intercept ACP initialization
+
+This ExecPlan (execution plan) is a living document. The sections
</code_context>
<issue_to_address>
**issue (review_instructions):** ACP is introduced in this file without being expanded on first use, which violates the acronym-definition requirement.

Please expand ACP on first use in this ExecPlan heading or immediately afterwards, for example: `Intercept Agentic Control Protocol (ACP) initialization`, and then use `ACP` thereafter.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: src/engine/connection/exec/protocol.rs [hunk 1]
```diff
@@ -177,11 +183,101 @@ where
...
 
+async fn forward_initial_protocol_frame_async<HostStdin>(
+    buffered_stdin: &mut tokio::io::BufReader<HostStdin>,
+    input: &mut Pin<Box<dyn AsyncWrite + Send>>,
+) -> io::Result<()>
+where
+    HostStdin: AsyncRead + Unpin,
+{
+    let mut first_frame = Vec::new();
+    if tokio::io::AsyncBufReadExt::read_until(buffered_stdin, b'\n', &mut first_frame).await? == 0 {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/protocol.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue.

### ❌ Failed checks (1 error, 7 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Module-Level Documentation | ❌ Error | PR introduces 12 undocumented public items across two files; Cargo.toml enforces missing\_docs = "deny" requiring /// documentation on all public constants and functions. | Add documentation comments to all 12 undocumented public items in acp\_helpers.rs and protocol\_acp\_tests.rs, then verify compilation with cargo build. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 62.75% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| User-Facing Documentation | ⚠️ Warning | User-facing documentation in docs/users-guide.md introduces ACP capability masking but fails to expand the acronym on first occurrence at line 66. | Update line 66 of docs/users-guide.md to expand ACP as Agentic Control Protocol (ACP) on first mention, aligning with established documentation standards. |
| Developer Documentation | ⚠️ Warning | The ACP acronym expansion is missing from developers-guide.md section 8.2.1 heading, which should read 'Agentic Control Protocol (ACP)' on first mention per coding guidelines. | Update the heading in developers-guide.md line 360 from '#### 8.2.1. ACP initialization rewrite contract' to '#### 8.2.1. Agentic Control Protocol (ACP) initialization rewrite contract'. |
| Testing (Unit) | ⚠️ Warning | BDD tests are disguised unit tests calling `run_forwarding()` with mocks rather than exercising the public protocol session API boundary. | Replace disguised BDD tests with genuine behavioural tests instantiating `ExecSessionOptions` with ACP rewrite enabled through `run_protocol_session_with_io_async`. |
| Testing (Property / Proof) | ⚠️ Warning | PR lacks property-based testing or bounded model checking for invariants spanning ranges of inputs, states, and transitions including buffer-boundary safety, frame delimitation, JSON transformation, and line-ending preservation. | Introduce proptest for arbitrary JSON frames with varying clientCapabilities structures and payloads; verify newline detection and frame delimitation across byte patterns up to MAX\_FIRST\_FRAME\_BYTES. |
| Domain Architecture | ⚠️ Warning | Protocol layer couples to ACP-specific logic; capability removal policy embedded in JSON serialisation without explicit domain abstraction; session layer exposed to protocol implementation details. | Extract CapabilityRestrictionPolicy domain abstraction; restructure acp\_helpers as pure JSON adapter; introduce pluggable frame-transformer interface; move rewrite\_acp\_initialize configuration to protocol layer. |
| Observability | ⚠️ Warning | PR introduces security-sensitive ACP capability masking without logging, tracing, or observability. Silent failures on JSON parse errors, missing ACP structure, and serialisation errors violate observability requirements already demonstrated in existing codebase. | Import and use the tracing crate to add structured logging at decision points with warn/debug statements for JSON failures, missing fields, and masking success; include operation name and diagnostic context. |
| Concurrency And State | ❓ Inconclusive | Concurrency model is explicit and provably correct with single-threaded per-session isolation, value-captured flags, local stack buffers, and explicit cancellation handling. However, unit tests do not directly exercise task cancellation during first-frame buffering or IO failure scenarios in isolation as required by the check. | Add unit tests for task abort during first-frame buffering, IO failure propagation, and output-loop shutdown racing with stdin buffering to verify safe cleanup under concurrent interleaving. |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/engine/connection/exec/protocol_acp_tests.rs (1)

`1-484`: _🧹 Nitpick_ | _🔵 Trivial_ | _⚖️ Poor tradeoff_
**Split this test module to stay within the 400-line limit.**
The file is 484 lines, exceeding the repository's 400-line budget. Extract the BDD scenario infrastructure (lines 364–483) into a sibling module such as `protocol_acp_bdd_tests.rs`, keeping the unit tests and forwarding tests in this file.

##### Proposed split

```rust
//! BDD scenarios for ACP capability masking.

use super::protocol_acp_tests::{
    AcpMaskingState, initialize_frame, initialize_without_blocked_capabilities,
    malformed_initialize_bytes, masked_initialize_with_follow_up, run_forwarding,
    session_new_bytes,
};
// ... move lines 364-483 here
```

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:

1. **Testing (Unit)** — Genuine BDD sessions via `ExecSessionOptions` + `run_protocol_session_with_io_async`
2. **Testing (Property/Proof)** — `proptest` for ACP frame invariants
3. **Domain Architecture** — `CapabilityRestrictionPolicy` trait extraction
4. **Concurrency And State** — Task-abort, IO-failure, and shutdown-race tests

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/engine/connection/exec/acp_helpers.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: parse_acp_frame_payload,serialize_acp_frame

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one has already been raised.

### ❌ Failed checks (2 errors, 6 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Test suite lacks rigorous coverage of independent capability removal and preservation of unrelated capabilities; all payloads contain only three clientCapabilities keys with no additional unrelated capabilities tested. | Add test cases with additional unrelated capabilities (auth, logging) to verify only fs and terminal are removed; add independent scenarios with only fs or only terminal present; include edge cases for frames without recognised line endings, oversized JSON frames, and nested structures; consolidate redundant BDD tests. |
| Unit Architecture | ❌ Error | Mixed responsibilities in async function; pure function silently hides JSON parse failures; observability side-effects not abstracted; error-handling paths not independently tested. | Refactor mask\_acp\_initialize\_frame to return Result<Vec, AcpMaskError>; extract frame detection into FrameDetector unit; inject AcpMaskingObserver trait for observability; add error-path unit tests. |
| Title check | ⚠️ Warning | The title omits the required roadmap reference; Step 2.6.1 must be included per guidelines. | Revise title to include the roadmap reference: 'Implement ACP capability masking at protocol proxy seam (2.6.1)'. |
| Developer Documentation | ⚠️ Warning | The pull request provides comprehensive documentation updates including an up-to-date execplan, marked-complete roadmap checkpoint, and enhanced developer's guide, but the ProtocolSessionOptions::with\_acp\_initialize\_rewrite\_enabled builder method in src/engine/connection/exec/protocol.rs lacks the required /// rustdoc comment violating section 15.3 guidelines. | Add a /// doc comment to the with\_acp\_initialize\_rewrite\_enabled method in ProtocolSessionOptions explaining its purpose, such as: /// Enable ACP initialisation rewriting for this protocol session. |
| Testing (Unit And Behavioural) | ⚠️ Warning | BDD tests violate custom check requirements by calling private `forward_host_stdin_to_exec_async` through mocks instead of exercising the public `run_protocol_session_with_io_async` API boundary. | Replace BDD tests with session-boundary integration tests calling `run_protocol_session_with_io_async` with real container streams, or integrate scenarios into `protocol_proxy_bdd.rs`. |
| Testing (Property / Proof) | ⚠️ Warning | Multiple substantive invariants across JSON serialisation, selective capability removal, line-ending preservation, bounded buffering, and state-machine transitions rely on discrete unit tests rather than property-based verification. | Introduce property-based tests using proptest to systematically verify invariants: arbitrary valid JSON generation, selective key removal assertions, line-ending variant testing, idempotence verification, and full size-range validation up to MAX\_FIRST\_FRAME\_BYTES. |
| Observability | ⚠️ Warning | Observability instrumentation lacks container context in logs, provides no metrics for frame masking operations, and omits trace spans across async boundaries. | Pass container ID through ACP helpers, add metrics for frame masking frequency and JSON failures, and instrument async boundaries with #\[instrument\] for production diagnostics. |
| Concurrency And State | ⚠️ Warning | The PR defers concurrency and cancellation safety testing to issue `#87`, leaving tokio::io::BufReader state unsafe across await points during task cancellation. | Document the concurrency model explicitly. Define ownership of the BufReader, document task cancellation and writer shutdown handling, ensure BufReader state cannot corrupt on cancellation, and add tests exercising task cancellation during reads and writer drops. |
| Domain Architecture | ❓ Inconclusive | No verification output provided to convert. | Supply the verification output that requires conversion into the specified JSON format. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

## New GitHub issue needed — enhanced Observability

The Observability check has escalated beyond the tracing already implemented: it now demands container ID threaded into log fields (requiring a signature change to `forward_initial_acp_frame_async`), metrics counters (new dependency), and `#[instrument]` on async boundaries. None of this is covered by issues `#84`–[#87](https://github.com/leynos/podbot/issues/87).

**Comment by leynos** [comment]
@coderabbitai resume

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/engine/connection/exec/protocol_acp_tests.rs: 

### What lead to degradation?
The module contains 5 functions with similar structure: initialize_frame,initialize_with_only_blocked_capabilities,mask_acp_initialize_frame_preserves_all_unrelated_capabilities,mask_acp_initialize_frame_removes_only_fs_when_terminal_absent and 1 more functions

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: src/engine/connection/exec/protocol_acp_tests.rs [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+    runtime
+        .block_on(async {
+            tokio::time::timeout(
+                STDIN_SETTLE_TIMEOUT,
+                forward_initial_acp_frame_async(&mut buffered_stdin, &mut container_input),
+            )
+            .await
+        })
+        .expect("initial forwarding should not wait for newline or EOF")
+        .expect("initial forwarding should succeed");
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/protocol_acp_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/engine/connection/exec/protocol_acp_tests.rs (1)

`1-523`: _🛠️ Refactor suggestion_ | _🟠 Major_ | _🏗️ Heavy lift_
**Split this test module below the 400-line budget.**
Bring this file back under the repository limit. It currently packs fixtures,

frame builders, forwarding harnesses, unit tests, shared BDD helpers, and the

BDD module include into 523 lines, which makes the ACP coverage harder to

maintain than the surrounding exec test layout.
Move the reusable frame builders and forwarding harness into a sibling helper

module, or move the exported BDD fixtures into the BDD file so this unit-test

module stays focused.
As per coding guidelines, "Files must not exceed 400 lines in length".

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests comprehensively cover mask\_acp\_initialize\_frame() behaviour and edge cases; forwarding tests exercise timeout and buffer boundaries. BDD tests merely wrap internal run\_forwarding() helper, duplicating unit test coverage without exercising public API. | Eliminate redundant BDD test wrapper layer. Add end-to-end tests exercising ExecSessionOptions.with\_acp\_initialize\_rewrite\_enabled() through full protocol session lifecycle, verifying flag propagation and masked initialisation frame delivery to container. |
| Domain Architecture | ⚠️ Warning | Domain logic for ACP capability masking directly manipulates serde\_json::Value and embeds JSON deserialisation/serialisation, violating segregation of domain logic from serialisation concerns. | Refactor to introduce adapter boundaries with pure domain functions operating on structured capability data, moving JSON handling into separate adapter functions. |
| Observability | ⚠️ Warning | ACP capability masking implementation lacks stable container context in logs, misaligned log levels for expected behaviour, and provides no metrics for feature enablement, masking success, or failure distinction. | Thread container\_id through ACP helpers via parameter or tracing span. Demote JSON parsing failures to debug level. Add structured logging at decision boundaries capturing rewrite enablement and frame modification status. Implement metrics for masking attempts and frame sizes. |
| Concurrency And State | ⚠️ Warning | Concurrency model is safe in practice but lacks explicit documentation and comprehensive testing of task cancellation. | Add documentation clarifying single-threaded execution semantics and add unit tests exercising task abortion mid-buffering. |

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | PR introduces new internal APIs and architectural changes, but the developers-guide repository layout is significantly outdated and does not document the new acp\_helpers.rs module or three ACP-specific test modules. | Update the repository layout diagram in docs/developers-guide.md section 3 to include acp\_helpers.rs and the three ACP test modules, add field-level doc comments to rewrite\_acp\_initialize in both structs explaining their purpose. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests validate `mask_acp_initialize_frame` comprehensively, but behavioural tests exercise test doubles instead of session-level protocol rewriting behaviour and integration. | Add integration test creating `ExecSessionOptions` with rewriting enabled, providing stdin with blocked ACP frame, and verifying masked output at session boundary in both enabled and disabled states. |
| Testing (Property / Proof) | ⚠️ Warning | The PR lacks property-based testing for mathematical invariants: bounded buffering, line-ending preservation, and selective capability removal across arbitrary inputs. | Introduce proptest for arbitrary byte sequences with random newlines, diverse JSON structures, and all line-ending variants; apply Kani bounded model checker to verify saturating subtraction arithmetic. |
| Domain Architecture | ⚠️ Warning | Domain logic for sandbox capability restrictions leaked into protocol adapter; business policy hardcoded as string constants in JSON manipulation code. | Extract a domain type representing SandboxCapabilityPolicy that explicitly lists denied capabilities; pass policy as parameter to ACP adapter rather than hardcoding. |
| Observability | ⚠️ Warning | ACP masking implementation lacks essential observability: missing container\_id context in logs, no metrics for operational visibility, no tracing spans for async task tracking, and no proper error categorisation for production diagnosis. | Thread container\_id through to acp\_helpers with stable contextual field; add #\[instrument(skip(...))\] for span creation; introduce metrics for masked frames, capacity events, and failures; escalate error handling to distinguish recoverable failures from expected pass-through behaviour. |
